domain/application/service/provider_test.go

```diff
@@ -1192,6 +1193,108 @@ func (s *providerServiceSuite) TestCreateApplicationWithStorageValidates(c *gc.C
 	c.Assert(err, gc.ErrorMatches, `.*invalid storage directives: charm "mine" has no store called "logs"`)
 }
 
+func (s *providerServiceSuite) TestDeviceConstraintsValidateNotInCharmMeta(c *gc.C) {
+	deviceConstraints := map[string]devices.Constraints{
+		"dev0": {
+			Type:  "type0",
+			Count: 42,
+		},
+	}
+	charmMeta := &charm.Meta{
+		Name: "foo",
+		Devices: map[string]charm.Device{
+			"dev1": {
+				Description: "dev1 description",
+				Type:        "type1",
+				CountMin:    1,
+			},
+		},
+	}
+
+	err := validateDeviceConstraints(deviceConstraints, charmMeta)
+	c.Assert(err, gc.ErrorMatches, "charm \"foo\" has no device called \"dev0\"")
+}
+
+func (s *providerServiceSuite) TestDeviceConstraintsValidateCount(c *gc.C) {
+	deviceConstraints := map[string]devices.Constraints{
+		"dev0": {
+			Type:  "type0",
+			Count: 42,
+		},
+	}
+	charmMeta := &charm.Meta{
+		Name: "foo",
+		Devices: map[string]charm.Device{
+			"dev0": {
+				Description: "dev0 description",
+				Type:        "type0",
+				CountMin:    43,
+			},
+		},
+	}
+
+	err := validateDeviceConstraints(deviceConstraints, charmMeta)
+	c.Assert(err, gc.ErrorMatches, "minimum device size is 43, 42 specified")
```

Comment:
I find it very weird there is a min, but not a max.

Reply:
Yeah, I had the same question, maybe @ycliuhw remembers?